src/utils/ucc_coll_utils.h

```diff
@@ -246,4 +250,49 @@ static inline size_t ucc_buffer_block_offset(size_t     total_count,
     return (block < left) ? offset - (left - block) : offset;
 }
 
+/* Given the rank space A (e.g. core ucc team), a subset B (e.g. active set
+   within the core team), the ep_map that maps ranks from the subset B to A,
+   and the rank of a process within A.
+   The function below computes the local rank of the process within subset B. */
+static inline ucc_rank_t ucc_ep_map_local_rank(ucc_ep_map_t map,
+                                               ucc_rank_t   rank)
+{
+    ucc_rank_t i, local_rank = UCC_RANK_INVALID;
+    int64_t    vrank;
+
+    switch(map.type) {
+    case UCC_EP_MAP_FULL:
+        local_rank = rank;
+        break;
+    case UCC_EP_MAP_STRIDED:
+        vrank = (int64_t)rank - (int64_t)map.strided.start;
+        vrank /= map.strided.stride;
+        ucc_assert(vrank >= 0 && vrank < map.ep_num);
+        local_rank = (ucc_rank_t)vrank;
+        break;
+    case UCC_EP_MAP_ARRAY:
+    case UCC_EP_MAP_CB:
+        for (i = 0; i < map.ep_num; i++) {
+            if (rank == ucc_ep_map_eval(map, i)) {
+                local_rank = i;
+                break;
+            }
+        }
+    default:
+        break;
+    }
+    return local_rank;
+}
```

Comment:
I think this code breaks if your stride is negative. For example:

```
0 Recv from 1 -> start = 1, stride = -1
1 Send to   0 -> start = 1, stride = -1
```

Similar problems when evaluating the rank in the map in `ucc_ep_map_eval`.

The simplest solution is probably to stash the team size in the map below so it can be referenced when needed.

Reply:
i don't see a problem with negative strides. In fact, it tried to add the support (and test in the gtest) for negative stride. Gtest covers those cases. Also i tried with ucc_sendrecv.cc from gist (check my comment there).
Regarding your example:
start = 1, stride - 1, so active set is {1, 0}
Then the code above would compute:
ucc_ep_map_local_rank for rank 0:
(0 - 1)/(-1) = 1
for rank 1:
(1 - 1)/(-1) = 0
looks correct

to do recv rank 0 posts bcast(active_set={1, 0}, root = 1); to do the corresponding send rank 1 posts bcast(active_set{1, 0}, root = 1);

The difference with my first proposal (where we had "my_ep" as part of active_set) is that now both ranks MUST provide the SAME active_set triplet.

subset of ranks {2,3} from a team of 5 ranks can be given 2 ways: start = 2, stride=1 OR start  = 3, stride - 1. In the former proposal we provided "my_ep"  (which implied the ordering of a set, ie local_set_ranks) and in that case it didn't matter if 2 procs would provided different definitions of active_set (2, 1) or (3, -1) - the correct local ranks were provided via my_ep. Now, we don't have it. The ordering is given by the start,stride It must be the same across active_set participants.

Reply:
I did find the problem. It came from me trying to back port these changes to an older internal version of this code. It's working now.